### ISSUE TITLE
MmFreePhysicalMemory race vs. GPU texture read

### ISSUE THREAD

**Issue by sabretooth** [benvanik/xenia#456]
One game that I tested (Legend of Spyro: Dawn of the Dragon) calls `MmAllocatePhysicalMemoryEx` to allocate GPU memory for UI textures, draws into those on the CPU side, references them in a GPU command buffer, issues the swap, and then frees the allocations at the start of the next frame with `MmFreePhysicalMemory`. This creates a race with the GPU thread to pull the data out of the texture allocation before the pages are marked no-access in `BaseHeap::Release`.

As a workaround to get the game booting, I changed the `xe::memory::Protect` call in `BaseHeap::Release` to mark the pages `kReadOnly` instead of `kNoAccess` so the GPU can still read them, but I'm not sure that this is a good long-term solution. Without documentation it's hard to know if it's a synchronization issue, if the GPU is supposed to keep access to physical allocations that the application has already released, or if this just happened to work with real hardware timings.

**Comment by benvanik** [comment]
Interesting! On the console the GPU has no concept of virtual memory and just references physical pages so it doesn't pay attention to the access protections, meaning that it'd be possible for this code to work without error on a real console. If it is intended to actually be doing that, though, it's stupid and dangerous - as if anything were to reallocate and reuse that memory you'd get corruption.

My guess would be there's something buggy with either a CPU or GPU-side wait/synchronize; like a fence not being waited on correctly. I do believe I've seen things that could be explained by that around swapping. You could try commenting out [this dispatch call](https://github.com/benvanik/xenia/blob/master/src/xenia/gpu/graphics_system.cc#L214) and seeing if it helps.

I don't have the game unfortunately, but if you could pastebin/etc up the logs while running it maybe I can see if something stands out on the CPU-side.

**Comment by DrChat** [comment]
There's a bunch of games that do this. Earthworm Jim is probably the best debug case as it's guaranteed to do this less than 5 seconds in.

**Comment by benvanik** [comment]
Perfect! I'll see if I can spot something.

**Comment by sabretooth** [comment]
Commenting out that dispatch call causes Spyro's main thread to lock up after a call to `VdSwap` + `VdRetrainEDRAM` before it can draw anything. I guess that technically prevents the crash, just not in a useful way.

**Comment by sawnikthederphog** [comment]
I remember xenia used to crash very inconsistently and frequently with The Legend of Spyro : Dawn of the Dragon, but on rare occasions it would make it to gameplay, nearly finished it actually. but after changes made to builds particularly around September, it has never gotten passed the splash screen. The weird manual cache clearing was present in the working builds. I've noticed in newer git builds that the clearing seems to work for the most part independently but i could be wrong. Many games crash in similar circumstances as well, Any ideas what changed? Anything I can do debug wise? After all this is my first time on Github.

Screenshots:

Menus/2D screens - unless buffer gets cleared manually it doesn't refresh however if it refreshes it crashes upon loading next FMV/texture. Normally freezes at start screen due to missing profile handling (bypassed using player 2 controller XD)

![xenia 05_12_2015 05_28_06 pm](https://cloud.githubusercontent.com/assets/16162330/11606692/e66285b6-9b75-11e5-85f5-de97916d9343.png)
![xenia 05_12_2015 05_28_24 pm](https://cloud.githubusercontent.com/assets/16162330/11606693/e9b0fd38-9b75-11e5-9d7a-2513cf67cb1e.png)

Ingame - geometry loads as per normal even with the buggy depth (probably fixed now) however alpha and transparent textures don't render correctly and the animation is...special


![xenia 05_12_2015 04_19_45 pm](https://cloud.githubusercontent.com/assets/16162330/11606650/5ccac508-9b74-11e5-9c4a-0f3b83ce5e29.png)


![xenia 05_12_2015 04_22_16 pm](https://cloud.githubusercontent.com/assets/16162330/11606662/b2b02fe4-9b74-11e5-8a19-17942e6c00e7.png)
![xenia 05_12_2015 04_22_57 pm](https://cloud.githubusercontent.com/assets/16162330/11606664/b5813916-9b74-11e5-80c7-6b3c0558d211.png)

Also awesome job on this emulator, the progress so far has been amazing, keep up the great work :)

**Comment by sabretooth** [comment]
The animation issues in that game should've been fixed in #457 and the hang on profile access should've been fixed in #455. Last I tried it (shortly after #457 was merged), with the workaround patch mentioned the first comment (changing the Protect call to mark the pages read-only instead of no-access), the game was playable. No music, though, and there was an issue with previous 2d content (menus and FMVs) getting merged into the background of whatever was rendering next until you hit the GPU cache clear button.

**Comment by sawnikthederphog** [comment]
Would i need to compile a build with your BaseHeap::Release patch? I would very much like to try this game out with fixed animations, I've never compiled xenia but I'll give it a try.

**Comment by sabretooth** [comment]
Yeah, without patching that line the game will crash before even making it to the title screen. src/xenia/memory.cc, line 817, change kNoAccess to kReadOnly.

**Comment by sawnikthederphog** [comment]
Thanks for the commit! This game looks beautiful now! Except now there is an odd depth issue with fluctuating characters, and its kinda scary.

![xenia 07_12_2015 04_31_53 pm](https://cloud.githubusercontent.com/assets/16162330/11620831/f5c7579c-9d01-11e5-937a-466894753378.png)
![xenia 07_12_2015 04_42_38 pm](https://cloud.githubusercontent.com/assets/16162330/11620843/0c6ecdfe-9d02-11e5-8592-e04f28acce7b.png)

![xenia 07_12_2015 04_42_38 pm](https://cloud.githubusercontent.com/assets/16162330/11620844/0f90cb72-9d02-11e5-9b3a-74723887896b.png)
![xenia 07_12_2015 04_32_04 pm](https://cloud.githubusercontent.com/assets/16162330/11620836/ff1b918c-9d01-11e5-95c1-3f8ca02b02da.png)

**Comment by DrChat** [comment]
This needs to be reviewed - this may have been fixed with recent changes to our texture handling code (we mistakenly thought textures took up more memory than they actually did).

**Comment by Oreos019** [comment]
![image](https://user-images.githubusercontent.com/34288678/33636219-32d71910-da13-11e7-8825-2ba9207e7edf.png)
